File: src/main/java/com/ahmedhossam/mnp/service/PortingRequestService.java
package com.ahmedhossam.mnp.service;

import com.ahmedhossam.mnp.dto.request.CreatePortingRequestDto;
import com.ahmedhossam.mnp.dto.response.PagedResponseDto;
import com.ahmedhossam.mnp.dto.response.PortingRequestResponseDto;
import com.ahmedhossam.mnp.enums.Operator;
import org.springframework.data.domain.Pageable;

public interface PortingRequestService {
    PortingRequestResponseDto create(CreatePortingRequestDto dto, Operator recipient);
    PortingRequestResponseDto accept(Long id, Operator caller);
    PortingRequestResponseDto reject(Long id, Operator caller);
    PagedResponseDto<PortingRequestResponseDto> list(Operator caller, Pageable pageable);
}

File: frontend/src/app/core/services/porting-request.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { CreatePortingRequestDto, PortingRequestResponseDto } from '../../models/porting-request.model';
import { PagedResponseDto } from '../../models/paged-response.model';
import { PhoneStatusResponseDto } from '../../models/phone-status.model';

@Injectable({ providedIn: 'root' })
export class PortingRequestService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/porting-requests`;
  private readonly phoneNumbersUrl = `${environment.apiUrl}/phone-numbers`;

  create(dto: CreatePortingRequestDto): Observable<PortingRequestResponseDto> {
    return this.http.post<PortingRequestResponseDto>(this.baseUrl, dto);
  }

  accept(id: number): Observable<PortingRequestResponseDto> {
    return this.http.post<PortingRequestResponseDto>(`${this.baseUrl}/${id}/accept`, {});
  }

  reject(id: number): Observable<PortingRequestResponseDto> {
    return this.http.post<PortingRequestResponseDto>(`${this.baseUrl}/${id}/reject`, {});
  }

  list(page: number = 0, size: number = 20): Observable<PagedResponseDto<PortingRequestResponseDto>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PagedResponseDto<PortingRequestResponseDto>>(this.baseUrl, { params });
  }

  getById(id: number): Observable<PortingRequestResponseDto> {
    return this.http.get<PortingRequestResponseDto>(`${this.baseUrl}/${id}`);
  }

  getPhoneStatus(phoneNumber: string): Observable<PhoneStatusResponseDto> {
    return this.http.get<PhoneStatusResponseDto>(`${this.phoneNumbersUrl}/${phoneNumber}/status`);
  }
}
